File: frontend/src/types/playlists.ts
import type { Track } from './track'

export interface PlaylistSummary {
  id: number
  name: string
  track_count: number
  created_at: string
  owner_username: string
  is_owner: boolean
  share_url: string | null
}

export interface PlaylistDetail {
  id: number
  name: string
  created_at: string
  owner_username: string
  is_owner: boolean
  share_url: string | null
  items: Track[]
}

export interface PlaylistListResponse {
  items: PlaylistSummary[]
  total: number
}

export interface SharedPlaylist {
  name: string
  owner_username: string
  tracks: Track[]
}

File: backend/app/schemas/playlists.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.tracks import TrackResponse


class PlaylistCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class PlaylistUpdateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class PlaylistResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    track_count: int
    created_at: datetime
    owner_username: str
    is_owner: bool
    share_url: str | None = None


class PlaylistDetailResponse(BaseModel):
    id: int
    name: str
    created_at: datetime
    owner_username: str
    is_owner: bool
    share_url: str | None = None
    items: list[TrackResponse] = []


class PlaylistListResponse(BaseModel):
    items: list[PlaylistResponse]
    total: int


class PlaylistAddTrackRequest(BaseModel):
    track_id: int = Field(ge=1)


class PlaylistOrderRequest(BaseModel):
    track_ids: list[int]


class SubscribeResponse(BaseModel):
    playlist_id: int


class SharedPlaylistResponse(BaseModel):
    name: str
    owner_username: str
    tracks: list[TrackResponse]
